### PR TITLE
Private/caolan/usermount5

### PR DESCRIPTION
**Pull request by caolanm**
* Resolves: # 
* Target version: master 

### TODO

- [ ] ...

### Checklist

- [ ] Code is properly formatted
- [ ] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

**Comment by caolanm** [comment]
still a WIP so not there yet,

These namespace things can't be created when there are threads running, and the Poco::Log rotateOnOpen option triggers one of these, via an archiver/zip-thing if it finds an existing log. I've just temp bodged that off here for simplicity[1].

mounts done by a child are local to it and so not visible to the parent, so the /tmp mounted in the child isn't seen at /path/to/jails/foo/jailid/tmp by the parent. If the parent uses where it is really mounted from of /path/to/jails/tmp/cool-jailid then things are straightforward. I've made this tweakery dependent on using mount-namespaces for now at least.

[1] One possibility might be if archive mode is used to open the log and close it again to force the archive to do its work and join that thread, losing presumably the neat ability to zip that in a thread while continuing to load. And then reopen the log again which probably doesn't then need to launch a thread.

File: wsd/Storage.cpp [hunk 1]
```diff
@@ -74,15 +74,44 @@ bool StorageBase::SSLEnabled = false;
...
     std::string localPath = _jailPath;
     if (localPath[0] == '/')
     {
         // Remove the leading /
         localPath.erase(0, 1);
     }
 
+    // Use where mountJail of kit/Kit.cpp mounts /tmp for this path *from* rather than
+    // where it is mounted *to*, so this process doesn't need the mount visible to it
+    if (COOLWSD::EnableMountNamespaces && !localPath.empty())
```
Comment (mmeeks) [inline wsd/Storage.cpp]: We should prolly do this un-conditionally in the next red release =)

File: kit/Kit.cpp [hunk 2]
```diff
@@ -3190,12 +3221,23 @@ void lokit_main(
             }
 
 #ifndef __FreeBSD__
-            dropCapability(CAP_SYS_CHROOT);
-            dropCapability(CAP_FOWNER);
-            dropCapability(CAP_CHOWN);
+            if (usingMountNamespace)
```
Comment (mmeeks) [inline kit/Kit.cpp:3224]: Cap dropping is tricky and seems to be copy/pasted below (?) =)

File: wsd/COOLWSD.cpp [hunk 3]
```diff
@@ -2560,9 +2552,40 @@ void COOLWSD::innerInitialize(Application& self)
 #endif // CODE_COVERAGE
 
     // Setup the jails.
+    const bool UseMountNamespaces = getConfigValue<bool>(conf, "mount_namespaces", false);
```
Comment (mmeeks) [inline wsd/COOLWSD.cpp:2555]: Bit of a surprise here; do we really need to do this inside coolwsd ? I'd expect this to be specific to the forkit; or are we containing the whole coolwsd now as well ? =) [ which makes good sense if I think about it too ;-]

**Review by mmeeks — APPROVED** [review]
Looks really good; a few comments below though =)

File: wsd/Storage.cpp (hunk 1 above)
Comment (caolanm) [reply wsd/Storage.cpp]: yeah, needs checking for existing fallback non-mount cases where I think we use link as to what links we are using and perhaps need to align the link fallback with the mount case so its always the same.

File: kit/Kit.cpp (hunk 2 above)
Comment (caolanm) [reply kit/Kit.cpp:3224]: can certainly bundle these together inside a utility

File: wsd/COOLWSD.cpp (hunk 3 above)
Comment (caolanm) [reply wsd/COOLWSD.cpp:2555]: In coolwsd we use setupChildRoot and that has a testing bind mount+umount of cool_test_mount to see if the full combination works and if it fails then disables bindmounting globally. So stuck that into the same type of combination of namespaces as used in the forkit.

Comment (mmeeks) [reply wsd/COOLWSD.cpp:2555]: Makes sense I guess =) So we're not changing coolwsd's view of the world / capabilities etc.

Quite possibly coolwsd should (after sucking in its file data) enter a namespace with no system file-system visibility either =) but that's for later - I wonder what best practice is there these days.

**Pull request merged**